core/src/main/java/org/apache/iceberg/DistributedDataBatchScan.java

```diff
@@ -0,0 +1,309 @@
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          if (planRemotely) {
+            scanMetrics().scannedDataManifests().increment(manifests.size());
+            List<DataFile> dataFiles = planDataRemotely(manifests);
+            int skippedDataFilesCount = totalFilesCount(manifests) - dataFiles.size();
+            scanMetrics().skippedDataFiles().increment(skippedDataFilesCount);
+            return CloseableIterable.withNoopClose(dataFiles);
+          } else {
+            ManifestGroup manifestGroup = newManifestGroup(manifests, ImmutableList.of());
+            return CloseableIterable.transform(manifestGroup.entries(), ManifestEntry::file);
+          }
+        },
+        workerPool);
+  }
+
+  private CompletableFuture<DeleteFileIndex> newDeletesFuture(
+      List<ManifestFile> manifests, boolean planRemotely, ExecutorService workerPool) {
+
+    return CompletableFuture.supplyAsync(
+        () -> {
+          DeleteFileIndex.Builder builder;
+
+          if (planRemotely) {
+            scanMetrics().scannedDeleteManifests().increment(manifests.size());
+            List<DeleteFile> deleteFiles = planDeletesRemotely(manifests);
+            int skippedDeleteFilesCount = totalFilesCount(manifests) - deleteFiles.size();
+            scanMetrics().skippedDeleteFiles().increment(skippedDeleteFilesCount);
+            builder = DeleteFileIndex.builderForFiles(io(), deleteFiles);
+          } else {
+            builder = DeleteFileIndex.builderFor(io(), manifests);
+          }
+
+          return builder
+              .specsById(table().specs())
+              .filterData(filter())
+              .caseSensitive(isCaseSensitive())
+              .planWith(planExecutor())
+              .scanMetrics(scanMetrics())
+              .build();
+        },
+        workerPool);
+  }
+
+  private CloseableIterable<ScanTask> planFilesLocally(
+      List<ManifestFile> dataManifests, List<ManifestFile> deleteManifests) {
+    ManifestGroup manifestGroup = newManifestGroup(dataManifests, deleteManifests);
+    return asScanTasks(manifestGroup.planFiles());
+  }
+
+  @Override
+  public CloseableIterable<ScanTaskGroup<ScanTask>> planTasks() {
+    return TableScanUtil.planTaskGroups(
+        planFiles(), targetSplitSize(), splitLookback(), splitOpenFileCost());
+  }
+
+  private CloseableIterable<ScanTask> createTasks(
+      CloseableIterable<DataFile> dataFiles, DeleteFileIndex deletes) {
+
+    String schemaString = SchemaParser.toJson(tableSchema());
+    LoadingCache<Integer, String> specStringCache = specCache(PartitionSpecParser::toJson);
+    LoadingCache<Integer, ResidualEvaluator> residualCache = specCache(this::newResidualEvaluator);
+
+    return CloseableIterable.transform(
+        dataFiles,
+        dataFile -> {
+          DeleteFile[] deleteFiles = deletes.forDataFile(dataFile);
```

Comment:
If this step could be done in executor that would be amazing. We have noticed some v2 tables blocking here.

Reply:
Could you, please, provide details on the use case stats? Size and number of data/delete files, type of delete files, compaction, etc.

Reply:
Hi @aokolnychyi, such as the following. The data is ingested into Iceberg with a Flink CDC job, and the following is a full table rewrite job for that. 
```
ScanMetricsResult{totalPlanningDuration=TimerResult{timeUnit=NANOSECONDS, totalDuration=PT45M35.528362027S, count=1}, resultDataFiles=CounterResult{unit=COUNT, value=4241121}, resultDeleteFiles=CounterResult{unit=COUNT, value=21491991}, totalDataManifests=CounterResult{unit=COUNT, value=138}, totalDeleteManifests=CounterResult{unit=COUNT, value=57}, scannedDataManifests=CounterResult{unit=COUNT, value=138}, skippedDataManifests=CounterResult{unit=COUNT, value=0}, totalFileSizeInBytes=CounterResult{unit=BYTES, value=70668583180805}, totalDeleteFileSizeInBytes=CounterResult{unit=BYTES, value=361014841132536}, skippedDataFiles=CounterResult{unit=COUNT, value=29}, skippedDeleteFiles=CounterResult{unit=COUNT, value=0}, scannedDeleteManifests=CounterResult{unit=COUNT, value=57}, skippedDeleteManifests=CounterResult{unit=COUNT, value=0}, indexedDeleteFiles=CounterResult{unit=COUNT, value=4181955}, equalityDeleteFiles=CounterResult{unit=COUNT, value=4181955}, positionalDeleteFiles=CounterResult{unit=COUNT, value=0}, skippedMetaBlobs=null}}

Reply:
Does anyone have any other thoughts on this or faced similar issues?

cc @RussellSpitzer @stevenzwu @jackye1995 @amogh-jahagirdar @rdblue @szehon-ho

Reply:
> 21 million delete files, 350 TB of deletes to apply

The same delete files can be applied to multiple data files. So the deleted file counting should be magnified.

> would this use case benefit from a regular job that would convert equality deletes into position deletes? 

This should be two jobs, first rewrite the equality deletes into position deletes and then apply the position deletes. That means the base file should be read twice. It could not provide many improvements compared with applying equality deletes directly. 

> Is there a time interval where that would be possible on a regular basis to avoid rewriting the entire set of data files? 

Hhhh, this is a fish that got away.

> Have you profiled it? Is most of the time spent on looking up the index?

This is a past job. We found there are no serial memory and IO problems from the logs. We may rerun the job and do some details profiling.

Reply:
> This should be two jobs, first rewrite the equality deletes into position deletes and then apply the position deletes. That means the base file should be read twice. It could not provide many improvements compared with applying equality deletes directly.

The problem with equality deletes is that they can't be rewritten across sequence numbers, essentially prohibiting a minor compaction. The number of equality deletes will increase until we do a major compaction (rewriting data is usually expensive). Position deletes, on the other hand, are much cheaper to apply on read and can be compacted across sequence numbers. On paper, an intermediate step to convert equality deletes into position deletes on a regular basis may be helpful as we will have a much smaller number of position delete files to apply until we do a major compaction and get rid of them. This is all theory that is based on the assumption that applying equality deletes and writing position deletes instead would be much cheaper than rewriting entire data files. One has to verify and prove it would actually be beneficial. I wonder if you are interested to give it a try given that you have a use case that suffers from a large number of equality deletes?

Reply:
@chenjunjiedada I remember you have done some work on the minor compaction.

Reply:
Maybe, we can do something for partition scoped deletes but may not be that easy if we have global deletes. One alternative that I quickly discarded was to wait until deletes are planned and then broadcast the delete index. I feel that would be a bad decision as the size of that index may not be trivial and there would be more back and forth communication and no way to plan data and deletes concurrently.

I will explore this a bit more.

Reply:
@ConeyLiu, there is also a performance issue in `DeleteFileIndex` that I discovered while benchmarking. It is related to min/max value conversion. It can be really slow... I've created #8157 to fix it.

Reply:
I think this is something we should skip for now. It's extremely helpful to take this step forward, without worrying about problems introduced by tables that are using a write strategy like Flink, where basically all work is deferred until later, and then is never done.

Reply:
I will measure the time it takes to perform this step on a table in a reasonable state. If it is not that bad, I'll probably skip for now. If it is a substantial bottleneck, we can consider a distributed assignment of deletes.

Reply:
After all optimizations for local delete assignment that went in, it is fairly cheap to assign position deletes. It is more expensive for equality deletes. I'd probably go with local assignment for now until we figure out the final vision for assignment of deletes.